# Lab1 - Scikit-learn
Author: Sam Rainbow

## 1. Introduction

The goal of this lab is to become familiar with the scikit-learn library.

You will practice loading example datasets, perform classification and regression with linear scikit-learn models, and investigate the effects of reducing the number of features (columns in X) and the number of samples (rows in X and y).


In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## 2. Classification

Using yellowbrick spam - classification  
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

The goal is to investigate `LogisticRegression(max_iter=2000)` and effects of reducing the number of features and number of samples on classification performance.

### 2.1 Implement convenience function

In [2]:
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

def get_classifier_accuracy(model, X, y):
    '''Calculate train and validation accuracy of classifier (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn classifier): Classifier to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training accuracy, validation accuracy
    
    '''
    
    #TODO: IMPLEMENT FUNCTION BODY
    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=956)
    model.fit(X_train, y_train)
    y_train_pred = model.predict(X_train)
    y_test_pred = model.predict(X_test)
    train_acc = accuracy_score(y_train, y_train_pred)
    val_acc = accuracy_score(y_test, y_test_pred)
    return train_acc, val_acc   

### 2.2 Load data

Use the yellowbrick function `load_spam()`, load the spam data set into feature matrix `X` and target vector `y`.

Print size and type of `X` and `y`.


In [20]:
# TODO: ADD YOUR CODE HERE
from yellowbrick.datasets import load_spam

X, y = load_spam()
print("Size of X: ", X.shape, " type: ", type(X))
print("Size of y: ", y.shape, " type: ", type(y))
print(X)



Size of X:  (4600, 57)  type:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (4600,)  type:  <class 'pandas.core.series.Series'>
      word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.21               0.28           0.50           0.0   
1               0.06               0.00           0.71           0.0   
2               0.00               0.00           0.00           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4595            0.31               0.00           0.62           0.0   
4596            0.00               0.00           0.00           0.0   
4597            0.30               0.00           0.30           0.0   
4598            0.96               0.00           0.00           0.0   
4599            0.00               0.00           0.65           0.0   

  

Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [12]:
# TODO: ADD YOUR CODE HERE
X_small, _, y_small, _ = train_test_split(X,y, train_size = 0.01, random_state=174)
print("Size of X: ", X_small.shape, " type: ", type(X))
print("Size of y: ", y_small.shape, " type: ", type(y))


Size of X:  (46, 57)  type:  <class 'pandas.core.frame.DataFrame'>
Size of y:  (46,)  type:  <class 'pandas.core.series.Series'>


### 2.3 Train and evaluate models

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
4. Call your convenience function `get_classifier_accuracy()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation accuracy for each call to the `results` DataFrame
6. Print `results`

In [25]:
# TODO: ADD YOUR CODE HERE
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(max_iter =2000)

col = ["Data Size", "Training Accuracy", "Validation Accuracy"]
results = pd.DataFrame(columns = col)

t_acc, v_acc = get_classifier_accuracy(model, X, y)
results.loc[0] = ["X and y",t_acc, v_acc]
t_acc, v_acc = get_classifier_accuracy(model, X.iloc[:,:2], y)
results.loc[1] = ["First 2 columns",t_acc, v_acc]
t_acc, v_acc = get_classifier_accuracy(model, X_small, y_small)
results.loc[2] = ["X and y small",t_acc, v_acc]

print(results)

         Data Size  Training Accuracy  Validation Accuracy
0          X and y           0.935072             0.917391
1  First 2 columns           0.608986             0.613043
2    X and y small           0.941176             0.750000


### 2.4 Questions
1. What is the validation accuracy using all data? What is the difference between training and validation accuracy?
2. How does the validation accuracy and difference between training and validation change when only two columns are used? Provide values.
3. How does the validation accuracy and difference between training and validation change when only 1% of the rows are used? Provide values.

1. The validation accuracy using all of the data is 0.917, with a difference of 0.018 between the training and validation accuracy. This indicates that model is accurately predicting unseen data.
2. When only two columns are used the validation accuracy decreases substantially. The difference between the training accuracy (0.609) and validation accuracy (0.613) is still low




## 3. Regression

Using yellowbrick energy - regression  
https://www.scikit-yb.org/en/latest/api/datasets/energy.html

The goal is to investigate `LinearRegression()` and effects of reducing the number of features and number of samples on regression performance.

### 3.1 Implement convenience function

In [ ]:
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split

def get_regressor_mse(model, X, y):
    '''Calculate train and validation mean-squared error (mse) of regressor (model)
        
        Splits feature matrix X and target vector y 
        with sklearn train_test_split() and random_state=956.
        
        model (sklearn regressor): Regressor to train and evaluate
        X (numpy.array or pandas.DataFrame): Feature matrix
        y (numpy.array or pandas.Series): Target vector
        
        returns: training mse, validation mse
    
    '''
   
    #TODO: IMPLEMENT FUNCTION BODY
    

### 3.2 Load data

Use the yellowbrick function `load_energy()` load the energy data set into feature matrix `X` and target vector `y`.

Print dimensions and type of `X` and `y`.

In [ ]:
# TODO: ADD YOUR CODE HERE


Using the sklearn function `train_test_split()` prepare a feature matrix `X_small` and target vector `y_small` that contain only **1%** of the rows. Use `random_state=174`.

Print size and type of `X_small` and `y_small`.

In [ ]:
# TODO: ADD YOUR CODE HERE


### 3.3 Train and evaluate models

1. Import `LinearRegression` from sklearn
2. Instantiate model `LinearRegression()`.
3. Create a pandas DataFrame `results` with columns: Data size, training MSE, validation MSE
4. Call your convenience function `get_regressor_mse()` using 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`
5. Add the data size, training and validation MSE for each call to the `results` DataFrame
6. Print `results`

In [ ]:
# TODO: ADD YOUR CODE HERE


### 3.4 Questions
1. What is the validation MSE using all data? What is the difference between training and validation MSE?
1. How does the validation MSE and difference between training and validation change when only two columns are used? Provide values.
1. How does the validation MSE and difference between training and validation change when only 1% of the rows are used? Provide values.

*YOUR ANSWERS HERE*



## 4. Observations/Interpretation

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*



## 5. Reflection
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*

